File: 백준/Gold/2493. 탑/탑.py
import sys
input = sys.stdin.readline

N = int(input())

towerHeights = list(map(int, input().split()))
stack = []
answer = [0] * N

for i in range(N):
  while stack:
    if stack[-1][1] > towerHeights[i]:
      answer[i] = stack[-1][0]
      break
    else:
      stack.pop()
  stack.append([i + 1, towerHeights[i]])

print(*answer)

File: 백준/Gold/2493. 탑/탑.js
const input = require('fs').readFileSync(0).toString().split('\n');

const N = input.shift();
const towerHeights = input[0].split(' ').map(Number);

const stack = [];
const answer = Array.from({ length: N }, () => 0);

towerHeights.forEach((height, index) => {
  while (stack.length !== 0) {
    const [targetIndex, targetHeight] = stack[stack.length - 1];

    if (targetHeight > height) {
      answer[index] = targetIndex;
      break;
    }

    stack.pop();
  }

  stack.push([index + 1, height]);
});

console.log(answer.join(' '));